### ISSUE TITLE
TypeError: prevDeps is undefined

### ISSUE THREAD

**Issue by metamn** [pbeshai/use-query-params#66]
Hello again,

I'm getting this strange error with the following code:

```js
  const { filtersURL, filters: defaultFilters } = props;

  const [filters, setFilters] = useState(defaultFilters);

  useEffect(() => {
    const fetchData = async () => {
      const result = await axios.get(filtersURL);

      if (result.data) {
        setFilters(result.data);
      }
    };
    fetchData();
  }, [filtersURL]);

  const queryParamsFromFilters = getQueryParamsFromFilters({
    filters: filters
  });

  const [queryParams, setQueryParams] = useQueryParams(queryParamsFromFilters);
```
`getQueryParamsFromFilters` returns a valid initialization object:

```js
test("Collects the query params from filters", () => {
  expect(getQueryParamsFromFilters({ filters: filters })).toStrictEqual({
    q: StringParam,
    location: ArrayParam
  });
});
```

The error message log referring to `useQueryParams` is:

```
useQueryParam
node_modules/use-query-params/esm/useQueryParam.js:33

  30 | 
  31 | 
  32 | var refHistory = React.useRef(history);
> 33 | React.useEffect(function () {
     | ^  34 |   refHistory.current = history;
  35 | }, [history]); // ref with current version location object (see #46)
  36 | 

useQueryParams/paramValues<
node_modules/use-query-params/esm/useQueryParams.js:63

  60 | 
  61 | var paramNames = Object.keys(paramConfigMap);
  62 | var paramValues = paramNames.map(function (paramName) {
> 63 |   return useQueryParam(paramName, paramConfigMap[paramName], rawQuery)[0];
     | ^  64 | }); // we use a memo here to prevent recreating the containing decodedValues object
  65 | // which would break === comparisons even if no values changed.
  66 | 

useQueryParams
node_modules/use-query-params/esm/useQueryParams.js:62

  59 | // we reuse the logic to not recreate objects
  60 | 
  61 | var paramNames = Object.keys(paramConfigMap);
> 62 | var paramValues = paramNames.map(function (paramName) {
     | ^  63 |   return useQueryParam(paramName, paramConfigMap[paramName], rawQuery)[0];
  64 | }); // we use a memo here to prevent recreating the containing decodedValues object
  65 | // which would break === comparisons even if no values changed.
```

And I also have a warning in the console:

```
Warning: React has detected a change in the order of Hooks called by Filters. This will lead to bugs and errors if not fixed. For more information, read the Rules of Hooks: https://fb.me/rules-of-hooks

   Previous render            Next render
   ------------------------------------------------------
1. useState                   useState
2. useEffect                  useEffect
3. useContext                 useContext
4. useRef                     useRef
5. useEffect                  useEffect
6. useMemo                    useMemo
7. useRef                     useRef
8. useEffect                  useEffect
9. useRef                     useRef
10. useEffect                 useEffect
11. useMemo                   useMemo
12. useContext                useContext
13. useRef                    useRef
14. useEffect                 useEffect
15. useRef                    useRef
16. useEffect                 useEffect
17. useMemo                   useMemo
18. useCallback               useCallback
19. useMemo                   useContext
   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

    in Filters (at App.js:11)
    in Route (created by QueryParamProvider)
    in QueryParamProvider (at App.js:10)
    in Router (created by BrowserRouter)
    in BrowserRouter (at App.js:9)
    in App (at src/index.js:6)
```

And I've saved a dedicated branch to help debug this issue: https://github.com/metamn/use-query-filters/tree/3-wrong-use-query-params
 
I hope it helps. I'm thinking that `useQueryParams` is not ready to handle async environments where the init params are not known a priori but are loaded from the database with a delay. Please see also https://github.com/pbeshai/use-query-params/issues/65

Thanks a lot!

**Comment by bmakan** [comment]
I had the same issue. I load parameters from the server and then call `useQueryParams`. What I ended up doing was wrapping my component in a "fetcher" component which would render `null` until data was fetched. Then the component calling `useQueryParams` would always have the config it requires.

At first I thought `useQueryParams` doesn't support empty object or changing the config object on the fly. The latter seems to be false, as I call it with different config based when an UI dropdown changes.

Code example:
```javascript
function Filters(props) {
    const {
        classes,
        dataType,
        filters,
    } = props;

    // Just trust me this yields a valid config - it can be different on each rerender
    const queryParamsConfig = _.reduce(_.map(filters[dataType], (filter) => filter.field), (config, param) => _.merge(config, { [param]: ArrayParam }), {});
    const [params, setParams] = useQueryParams(queryParamsConfig);

    return filters[dataType].map((filter) => (
        <div className={classes.multiselect} key={filter.field}>
            <MultiSelect
                defaultValues={params[filter.field]}
                label={filter.label}
                multiselect
                onChange={setParams[filter.field]}
                options={filter.options}
            />
        </div>
    ));
}

function SpecificFilters(props) {
    const {
        classes,
        dataType,
        fetchSpecificFilters: fetchSpecificFiltersFunc,
        filters,
    } = props;

    // Fetch the data and ensure it's present
    useEffect(() => {
        fetchSpecificFiltersFunc();
    }, [fetchSpecificFiltersFunc]);

    if (!filters) {
        return null;
    }
    if (!filters[dataType]) {
        return null;
    }

    // Only render when API fetch is complete
    return <Filters classes={classes} dataType={dataType} filters={filters} />;
}
```

**Comment by dprgarner** [comment]
I think this issue is because the library doesn't currently support dynamic configs in `useQueryParams`, as there's code breaking the rules of hooks. [I've raised a comment on this issue](https://github.com/pbeshai/use-query-params/issues/72#issuecomment-614223009). 

@bmakan are you sure that your code is working with dynamic configs? When investigating, I found that dynamic configs would break when the number of query params changed. Perhaps your reducer is always in practice returning a config with a single array param? Or is the `MultiSelect` prop `onChange` not quite right?
